pkg/streamingpromql/engine_fuzz_test.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/promqltest"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	types.EnableManglingReturnedSlices = true
+	parser.ExperimentalDurationExpr = true
+	parser.EnableExperimentalFunctions = true
+}
+
+// we ignore these as they include the string index of where the invalid value is in the query
+// seems like prometheus does a trim() somewhere but MQE does not, so we end up with different char index's referenced in the warning
+var ignorePrefix = []string{"PromQL warning: quantile value should be between 0 and 1", "PromQL info: metric might not be a counter, name does not end in"}
+
+// Replace series Metric which are nil with an empty Labels{}
+func fixUpEmptyLabels(r *promql.Result) {
+	if r == nil || r.Value == nil {
+		return
+	}
+	switch r.Value.Type() {
+	case parser.ValueTypeMatrix:
+		matrix, _ := r.Matrix()
+		for i := 0; i < len(matrix); i++ {
+			if matrix[i].Metric == nil {
+				matrix[i].Metric = labels.Labels{}
+			}
+		}
+	}
+}
+
+func DoComparisonQuery(t *testing.T, asInstantQuery bool, query string, prometheusEngine *promql.Engine, mimirEngine *Engine, testStorage *teststorage.TestStorage) {
+
+	var prom, mimir *promql.Result
+	var promErr, mimiErr error
+	var promQry, mimiQry promql.Query
+
+	if asInstantQuery {
+
+		ts := time.Now()
+
+		promQry, promErr = prometheusEngine.NewInstantQuery(context.Background(), testStorage, nil, query, ts)
+		mimiQry, mimiErr = mimirEngine.NewInstantQuery(context.Background(), testStorage, nil, query, ts)
+
+	} else {
+
+		start := time.Now()
+		end := time.Now().Add(time.Minute)
+		step := time.Minute
+
+		promQry, promErr = prometheusEngine.NewRangeQuery(context.Background(), testStorage, nil, query, start, end, step)
+		mimiQry, mimiErr = mimirEngine.NewRangeQuery(context.Background(), testStorage, nil, query, start, end, step)
+	}
+
+	if promErr == nil {
+		defer promQry.Close()
+		prom = promQry.Exec(context.Background())
+	}
+
+	if mimiErr == nil {
+		defer mimiQry.Close()
+		mimir = mimiQry.Exec(context.Background())
+	}
+
+	if promErr == nil {
+
+		// prometheus may have a metric label = nil, and mimir could have an empty slice of metric labels
+		fixUpEmptyLabels(prom)
+
+		skipAnnotationComparison := false
+		if prom != nil && mimir != nil && prom.Warnings != nil {
+			for e := range prom.Warnings {
+
+				for _, prefix := range ignorePrefix {
+					if strings.HasPrefix(e, prefix) {
+						skipAnnotationComparison = true
+						break
+					}
+				}
+			}
+		}
+
+		if mimir != nil && prom != nil && prom.Err != nil && prom.Err.Error() == "expanding series: unexpected all postings" {
+			assert.Equal(t, "unexpected all postings", mimir.Err.Error())
+		} else {
+			testutils.RequireEqualResults(t, query, prom, mimir, skipAnnotationComparison)
+		}
+	} else {
+		require.NotNil(t, mimiErr)
+	}
+}
+
+// This is a utility test which can be useful testing individual query strings
```

Comment:
Is this needed? It should be possible to run the fuzz test with a filter to specify a single expression to run (or does that not work with fuzz tests?)